dependency/retrieval/main.go

```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"flag"
+	"log"
+	"os"
+
+	"github.com/paketo-buildpacks/dotnet-core-sdk/dependency/retrieval/components"
+	"github.com/paketo-buildpacks/packit/v2/cargo"
+)
+
+func main() {
+	var buildpackTOMLPath, outputPath string
+	set := flag.NewFlagSet("", flag.ContinueOnError)
+	set.StringVar(&buildpackTOMLPath, "buildpack-toml-path", "", "path to the buildpack.toml file")
+	set.StringVar(&outputPath, "output", "", "path to the output file")
+	err := set.Parse(os.Args[1:])
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fetcher := components.NewFetcher()
+	releases, err := fetcher.Get()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var versions []string
+	for _, release := range releases {
+		versions = append(versions, release.SemVer.String())
+	}
+
+	newVersions, err := components.FindNewVersions(buildpackTOMLPath, versions)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var dependencies []cargo.ConfigMetadataDependency
+	for _, version := range newVersions {
+		for _, r := range releases {
+			if r.SemVer.String() == version {
+				dependency, err := components.ConvertReleaseToDependency(r)
```

Comment:
I think the issue is that you are converting ALL `newVersions` includes too many versions, so you're converting SO many releases into dependencies.

Instead, I think `newVersions` should be whittled down in the `FindNewVersions` command to ONLY return as many new versions as are needed by the `buildpack.toml` constraints.

Reply:
This feels like we are essentially offloading the `jam` replacement logic on to the retrieval action for what is a relatively niche case (i.e. there are no existing versions in the buildpack and a constraint has just been added where there are a lot of matching versions). When doing just 1 or 2 versions the run time is much more acceptable so I am not sure that this optimization is worth implementing but I would be interested to hear further thoughts on this.

Reply:
yes this and `jam` will do similar things in terms of filtering versions down. It would be great if this functionality was factored out somewhere for everyone to use. 

Anyways, the way I have explained it should work is how its outlined in the RFC. The reason we need to do it like this is so that everything will run smoothly in the workflows, of which we have just 1 for all dependency cases. In cases where the dependency needs to be compiled, it uses the list of versions in the metadata.json file to figure out which versions need to be compiled. So in that case, the retrieval code NEEDs to just return the versions we definitely want otherwise we will needlessly compile extra versions.

Reply:
I see your point that not doing this here will most likely work because the versions will get fed into `jam update-dependencies` and get filtered. However, I think I'd still like the code to behave in the expected way for 2 reasons:
- uniformity. I'd love for retrieval to perfom the same way in every buildpack, rather than returning all versions for some buildpacks, and just the constraint-matching dependencies in other cases
- performance. I realize it introduces more code, but it will greatly reduce the runtime of this code, which is ideal. It's less than ideal to have the retrieval step run for 5-10 minutes everytime we poll for new versions.

Reply:
Ah I see what you mean that in most cases, it will only pick up one or two newer versions so it won't be a huge performance hit

Reply:
That is a reasonable explanation. I will update the code really quickly.

Reply:
I have added the constraint checking code to the latest version. Feel free to check it out.